File: cpp/leetcode/leetcode/10_regular_expression_matching.cpp
//
//  10_regular_expression_matching.cpp
//  leetcode
//
//  Created by hatlonely on 16/1/11.
//  Copyright © 2016年 hatlonely. All rights reserved.
//
//  Implement regular expression matching with support for '.' and '*'.
//
//  '.' Matches any single character.
//  '*' Matches zero or more of the preceding element.
//
//  The matching should cover the entire input string (not partial).
//
//  The function prototype should be:
//  bool isMatch(const char *s, const char *p)
//
//  Some examples:
//  isMatch("aa","a") → false
//  isMatch("aa","aa") → true
//  isMatch("aaa","aa") → false
//  isMatch("aa", "a*") → true
//  isMatch("aa", ".*") → true
//  isMatch("ab", ".*") → true
//  isMatch("caab", "c*a*b") → true
//

#include <iostream>
#include <cassert>
#include <string>

namespace regular_expression_matching {

class Solution {
public:
    bool isMatch(std::string source, std::string pattern) {
        int lens = (int)source.length();
        int lenp = (int)pattern.length();
        
        int idxs = 0;
        int idxp = 0;
        
        while (idxs < lens && idxp < lenp) {
            if (idxp + 1 < lenp && pattern[idxp + 1] == '*') {
                int idx = idxp++;
                while (idxp < lenp && pattern[idxp] == '*') {
                    idxp++;
                }
                while ((pattern[idx] == '.' || pattern[idx] == source[idxs])) {
                    if (!isMatch(source.substr(idxs), pattern.substr(idxp))) {
                        idxs++;
                        if (idxs > lens) {
                            return false;
                        }
                    } else {
                        return true;
                    }
                }
            } else {
                if (pattern[idxp] == '.') {
                    idxs++;
                    idxp++;
                } else if (pattern[idxp] == source[idxs]) {
                    idxs++;
                    idxp++;
                } else {
                    return false;
                }
            }
        }
        
        if (idxs == lens) {
            while (idxp + 1 < lenp && pattern[idxp + 1] == '*') {
                idxp++;
                while (idxp < lenp && pattern[idxp] == '*') {
                    idxp++;
                }
                if (idxp == lenp) {
                    return true;
                }
            }
            return idxp == lenp;
        } else {
            return false;
        }
    }
};

int main(int argc, const char *argv[]) {
    Solution solution;
    
    assert(!solution.isMatch("aa","a"));
    assert(!solution.isMatch("aaa","aa"));
    assert(!solution.isMatch("", "c*ab"));
    
    assert(solution.isMatch("aa","aa"));
    assert(solution.isMatch("aa", "a*"));
    assert(solution.isMatch("aa", ".*"));
    assert(solution.isMatch("aab", "c*a*b"));
    assert(solution.isMatch("a", "ab*"));
    
    return 0;
}
}

File: go/src/main/10_regular_expression_matching.go
package main

import "fmt"

func isMatch(source string, pattern string) (match bool) {
	if len(source) != 0 && len(pattern) == 0 {
		return false
	}
	if len(source) == 0 && len(pattern) == 0 {
		return true
	}

	star := len(pattern) >= 2 && pattern[1] == '*'

	if len(source) > 0 && (source[0] == pattern[0] || pattern[0] == '.') {
		if star {
			return isMatch(source, pattern[2:]) || isMatch(source[1:], pattern)
		} else {
			return isMatch(source[1:], pattern[1:])
		}
	} else {
		if star {
			return isMatch(source, pattern[2:])
		} else {
			return false
		}
	}

	return false
}

func main() {
	for _, unit := range []struct {
		source  string
		pattern string
		match   bool
	}{
		{"aa", "a", false},
		{"aa", "aa", true},
		{"aaa", "aa", false},
		{"aa", "a*", true},
		{"aa", ".a", true},
		{"ab", ".*", true},
		{"aab", "c*a*b", true},
		{"a", "ab*", true},
	} {
		if match := isMatch(unit.source, unit.pattern); match != unit.match {
			fmt.Printf("isMatch(%v, %v) = %v != %v\n", unit.source, unit.pattern, match, unit.match)
		}
	}
}
